include/wil/registry_helpers.h

```diff
@@ -0,0 +1,979 @@
+                {
+                    const auto error = ::RegSetKeyValueW(
+                        m_key,
+                        subkey,
+                        value_name,
+                        type,
+                        static_cast<BYTE*>(reg_value_type_info::get_buffer(value)),
+                        reg_value_type_info::get_buffer_size(value));
+                    return err_policy::HResult(HRESULT_FROM_WIN32(error));
+                }
+
+                template <typename R, typename get_value_with_type_policy = err_policy>
+                typename get_value_with_type_policy::result get_value_with_type(_In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, R& return_value, DWORD type = reg_value_type_info::get_value_type<R>()) const
+                {
+                    HRESULT hr = S_OK;
+
+                    if
+#if defined(__cpp_if_constexpr)
+                        constexpr
+#endif
+                        (reg_value_type_info::supports_prepare_buffer<R>())
+
+                    {
+                        hr = reg_value_type_info::prepare_buffer(return_value);
+                        if (FAILED(hr))
+                        {
+                            return get_value_with_type_policy::HResult(hr);
+                        }
+                    }
+
+                    DWORD bytes_allocated{ reg_value_type_info::get_buffer_size(return_value) };
+                    for (;;)
+                    {
+                        constexpr DWORD* null_type{ nullptr };
+                        DWORD data_size_bytes{ bytes_allocated };
+                        hr = HRESULT_FROM_WIN32(::RegGetValueW(
+                            m_key,
+                            subkey,
+                            value_name,
+                            get_value_flags_from_value_type(type),
+                            null_type,
+                            reg_value_type_info::get_buffer(return_value),
+                            &data_size_bytes));
+
+                        // some return types we can grow as needed - e.g. when writing to a std::wstring
+                        // will only compile the below for those types that support dynamically growing the buffer
+                        if
+#if defined(__cpp_if_constexpr)
+                            constexpr
+#endif
+                            (reg_value_type_info::supports_resize_buffer<R>())
+                        {
+                            // GetRegValueW will indicate the caller allocate the returned number of bytes in one of two cases:
+                            // 1. returns ERROR_MORE_DATA
+                            // 2. returns ERROR_SUCCESS when we gave it a nullptr for the out buffer
+                            const bool shouldReallocate = (hr == HRESULT_FROM_WIN32(ERROR_MORE_DATA)) || (SUCCEEDED(hr) && (reg_value_type_info::get_buffer(return_value) == nullptr) && (data_size_bytes > 0));

```

Comment:
Is the ERROR_MORE_DATA check a helper function in `registry.h`? Can we use it here?

Reply:
yep. thanks!